src/main/java/fi/aalto/cs/apluscourses/model/APlusAuthentication.java

```diff
@@ -0,0 +1,26 @@
+package fi.aalto.cs.apluscourses.model;
+
+import java.util.Arrays;
+import org.jetbrains.annotations.NotNull;
+
+public class APlusAuthentication {
+  @NotNull
+  private char[] token;
+
+  /**
+   * Initializes an instance with the given token. Note, that the given token array is not cleared
+   * or overwritten.
+   */
+  public APlusAuthentication(@NotNull char[] token) {
+    this.token = token.clone();
+  }
+
+  @NotNull
+  public char[] getToken() {
+    return token;
+  }
+
+  public void clear() {
+    Arrays.fill(token, '\0');
+  }
```

Comment:
Have you designed who calls this and when?

Reply:
Not sure yet. As mentioned in another comment to Nikolai, we'll probably end up having the token in memory whenever the IDE is running (and the token has been loaded), so it's not easy to decide where this would get called.

Anyways the utility of this is somewhat limited, since the GC may have moved the array at some point, potentially leaving the token in plaintext somewhere else in memory, which we can't overwrite. Also, if someone is actually reading the memory, then a breach has already occurred. If we really want to play it safe, we should probably use something like [Oracle's GuardedString](https://docs.oracle.com/html/E28160_01/org/identityconnectors/common/security/GuardedString.html), which encrypts its contents using a random key, the plaintext contents are accessed using callbacks, and the arrays are overwritten always after a callback is done.

Reply:
I looked at that `GuardedString` too but I guess it may be an overkill here. After all, we have no nuclear missile launch codes here. I guess we can achieve with carefully used `char[]` token a security level in which the effort to crack the system overly exceeds the gain an attacker could possibly achieve.

Bigger risk than someone maliciously getting access to the token in some GC-abandoned memory area, is, IMO, the possibility that some of us (or someone maintaining the software) accidentally end up writing code that prints the token to some log dump or somewhere else where it doesn't belong. So I'd prefer such design that minimizes the chance for that to happen.